### ISSUE TITLE
Image data augmentation in BatchIterator

### ISSUE THREAD

**Issue by FlorianMuellerklein** [dnouri/nolearn#70]
First, thank you very much for this. It's a fantastic tool!

I'm sorry to create an issue if this is user error but I'm totally stumped. Doing some googling, I've also saw that someone had a similar problem in the comments on the facial keypoints detection and I've seen comments about this same issue in the kaggle data science bowl competition. So I know that I'm not alone. 

Anyway, I'm trying to modify the FlipBatchIterator example from your kfkd example to do some image translations (rotation and translation). Everything starts out ok and then after 10 epochs or so both the train and test loss starts to go up and the validation accuracy goes down. It could very well be a bug in my code but I can't figure it out. (This is MNIST data)

Here is the code. If I uncomment the plot_sample function and look at each image as it goes through the batch iterator it looks perfectly fine as well. 

```python
class DataAug(BatchIterator):

    def transform(self, Xb, yb):
        Xb, yb = super(DataAug, self).transform(Xb, yb)
        
        for i in range(Xb.shape[0]):    
            # random rotations
            dorotate = randint(-3,3)
    
            # random translation
            trans_1 = randint(-2,2)
            trans_2 = randint(-2,2)
    
            # random zoom
            #zoom_1 = uniform(1, 1.01)
    	
            tform = transform.AffineTransform(translation = (trans_1, trans_2)) 
									#scale = (1/zoom_1, 1/zoom_1))
            
                        
            new = transform.warp(Xb[i][0], tform, 
                            output_shape = (PIXELS, PIXELS),
                            mode = 'constant').astype(np.float32)
                        
            new = transform.rotate(image = new, angle = dorotate, 
                            resize = False, mode = 'constant').astype(np.float32)

            Xb[i][0] = new
            #plot_sample(Xb[i])
    
        return Xb, yb
```
And here is an image of the plot with the batchiterator loss.
![mnist_aug_withbatch](https://cloud.githubusercontent.com/assets/6733376/7223119/5918269c-e6ea-11e4-89f4-f8de798620ca.png)

Here is a plot in which I expanded the size of the dataset by doing the transformations and then appending them to the dataset. This worked very well, even though it was extremely slow due to the dataset size. 
![mnist_aug_notwithbatch](https://cloud.githubusercontent.com/assets/6733376/7223128/780d5658-e6ea-11e4-9a2f-a8f5d99ab0ca.png)

Am I doing something stupid or is this an issue? Thank you again for this great tool.

**Comment by dnouri** [comment]
When you write into `Xb` those changes will be persisted.  So next time you get an `Xb`, it will have images that are already translated and rotated, and you do it again.  If this is what's happening, then the way to fix this is to return a copy of `Xb` instead of writing into it.

**Comment by FlorianMuellerklein** [comment]
Oh wow, so I was doing something stupid. Thank you very much, that fixed it.
